File: pre-process/src/main/scala/com/oracle/ugbu/ouai/ad/utils/SparkSessionUtils.scala
package com.oracle.ugbu.ouai.ad.utils

import com.oracle.ugbu.ouai.ad.constants.OCIConstants
import com.typesafe.config.ConfigFactory
import org.apache.spark.SparkConf
import org.apache.spark.sql.SparkSession
import org.slf4j.{Logger, LoggerFactory}

import java.io.File
import java.util.TimeZone

object SparkSessionUtils {

  val logger: Logger = LoggerFactory.getLogger("AppUtils")

  def createSparkSession(configFilePath: String): SparkSession = {
    logger.info("---------->>> Create OCI SparkSession <<<----------")

    TimeZone.setDefault(TimeZone.getTimeZone("UTC"))
    val osConfig = ConfigFactory.parseFile(new File(configFilePath))

    logger.info("---->>> OS config object is -->> {}", osConfig.entrySet())
    val conf = new SparkConf()
      .setAppName("SpeedRacer")
      .set(
        "fs.oci.client.hostname",
        osConfig.getString(OCIConstants.OS_HOSTNAME)
      )
      .set(
        "fs.oci.client.auth.tenantId",
        osConfig.getString(OCIConstants.TENANCY_ID)
      )
      .set(
        "fs.oci.client.auth.userId",
        osConfig.getString(OCIConstants.USER_ID)
      )
      .set(
        "fs.oci.client.auth.fingerprint",
        osConfig.getString(OCIConstants.FINGERPRINT)
      )
      .set(
        "fs.oci.client.auth.pemfilepath",
        osConfig.getString(OCIConstants.PEM_FILE_PATH)
      )
      .set("fs.bmc.impl", "com.oracle.bmc.hdfs.BmcFilesystem")
      .set(
        "spark.delta.logStore.class",
        "org.apache.spark.sql.delta.storage.OciSingleDriverLogStore"
      )
      .set("spark.databricks.delta.schema.autoMerge.enabled", "true")

    val sparkSsn = SparkSession.builder().config(conf).getOrCreate()

    sparkSsn.sparkContext.hadoopConfiguration.set(
      "fs.oci.client.hostname",
      osConfig.getString(OCIConstants.OS_HOSTNAME)
    )
    sparkSsn.sparkContext.hadoopConfiguration.set(
      "fs.oci.client.auth.tenantId",
      osConfig.getString(OCIConstants.TENANCY_ID)
    )
    sparkSsn.sparkContext.hadoopConfiguration.set(
      "fs.oci.client.auth.userId",
      osConfig.getString(OCIConstants.USER_ID)
    )
    sparkSsn.sparkContext.hadoopConfiguration.set(
      "fs.oci.client.auth.fingerprint",
      osConfig.getString(OCIConstants.FINGERPRINT)
    )
    sparkSsn.sparkContext.hadoopConfiguration.set(
      "fs.oci.client.auth.pemfilepath",
      osConfig.getString(OCIConstants.PEM_FILE_PATH)
    )
    sparkSsn.sparkContext.hadoopConfiguration
      .set("namespace", osConfig.getString(OCIConstants.NAMESPACE))

    // Set shuffle partition - no. of executors * no.of cores per executor
    sparkSsn.conf
      .set(
        "spark.sql.shuffle.partitions",
        (sparkSsn.conf
          .get("spark.executor.instances")
          .toInt * sparkSsn.conf
          .get("spark.kubernetes.executor.limit.cores")
          .toInt).toString
      )

    logger.info("---->>> Spark config object is -->> {}", sparkSsn.conf.getAll)

    logger.info(
      "------------------>>> SPARK SESSION CREATED <<<------------------"
    )
    sparkSsn
  }

}

File: scoring/disagg_insights/utils/spark_session_utils.py
"""
spark_session_utils.py
~~~~~~~~

Module containing helper function for use with Apache Spark
"""
from pyhocon import ConfigList
from pyspark.sql import SparkSession

from disagg_insights.constants import OCIConstants


def create_spark_session(os_config_parsed: ConfigList) -> SparkSession:
    """
    :param os_config_parsed:
    :return:
    """
    spark_ssn = (SparkSession.builder
                 .appName("SpeedRacer")
                 .config("fs.oci.client.hostname",
                         os_config_parsed.get_string(OCIConstants['OS_HOSTNAME']))
                 .config("fs.oci.client.auth.tenantId",
                         os_config_parsed.get_string(OCIConstants['TENANCY_ID']))
                 .config("fs.oci.client.auth.userId",
                         os_config_parsed.get_string(OCIConstants['USER_ID']))
                 .config("fs.oci.client.auth.fingerprint",
                         os_config_parsed.get_string(OCIConstants['FINGERPRINT']))
                 .config("fs.oci.client.auth.pemfilepath",
                         os_config_parsed.get_string(OCIConstants['PEM_FILE_PATH']))
                 .config("fs.bmc.impl",
                         "com.oracle.bmc.hdfs.BmcFilesystem")
                 # .config("spark.delta.logStore.class",
                 #         "org.apache.spark.sql.delta.storage.OciSingleDriverLogStore")
                 # when running on local only
                 .config("spark.delta.logStore.oci.impl",
                         "io.delta.storage.OracleCloudLogStore")
                 .config("spark.databricks.delta.schema.autoMerge.enabled", "true")
                 # when running on local only
                 # .config("spark.sql.autoBroadcastJoinThreshold", "-1")
                 .config("spark.sql.broadcastTimeout", "10000")
                 .getOrCreate())

    spark_ssn.sparkContext._jsc.hadoopConfiguration().set(
        "fs.oci.client.hostname", os_config_parsed.get_string(OCIConstants['OS_HOSTNAME']))

    spark_ssn.sparkContext._jsc.hadoopConfiguration().set(
        "fs.oci.client.auth.tenantId", os_config_parsed.get_string(OCIConstants['TENANCY_ID']))

    spark_ssn.sparkContext._jsc.hadoopConfiguration().set(
        "fs.oci.client.auth.userId", os_config_parsed.get_string(OCIConstants['USER_ID']))

    spark_ssn.sparkContext._jsc.hadoopConfiguration().set(
        "fs.oci.client.auth.fingerprint", os_config_parsed.get_string(OCIConstants['FINGERPRINT']))

    spark_ssn.sparkContext._jsc.hadoopConfiguration().set(
        "fs.oci.client.auth.pemfilepath", os_config_parsed.get_string(OCIConstants['PEM_FILE_PATH']))

    spark_ssn.sparkContext._jsc.hadoopConfiguration().set(
        "namespace", os_config_parsed.get_string(OCIConstants['NAMESPACE']))

    # uncomment when running on cluster
    # (spark_ssn.sparkContext.getConf().set(
    #     "spark.sql.shuffle.partitions",
    #     ([int(spark_ssn.conf.get("spark.executor.instances"))
    #       if 'spark.executor.instances' in
    #          [i[0] for i in spark_ssn.sparkContext.getConf().getAll()] else 1][0] *
    #      [int(spark_ssn.conf.get("spark.kubernetes.executor.limit.cores"))
    #       if 'spark.kubernetes.executor.limit.cores' in
    #          [i[0] for i in spark_ssn.sparkContext.getConf().getAll()] else 1][0]
    #      )
    # ))

    spark_ssn.sparkContext.setLogLevel("error")
    return spark_ssn
